kernel/yosys.h

```diff
@@ -305,8 +305,8 @@ RTLIL::IdString new_id(std::string file, int line, std::string func);
 #define NEW_ID \
 	YOSYS_NAMESPACE_PREFIX new_id(__FILE__, __LINE__, __FUNCTION__)
 
-#define ID(_str) \
-	([]() { static YOSYS_NAMESPACE_PREFIX RTLIL::IdString _id(_str); return _id; })()
+#define ID(_id) ([]() { const char *p = "\\" #_id, *q = p[1] == '$' ? p+1 : p; \
```

Comment:
Can you add a brief comment here explaining the purpose/reason for this macro (i.e. performance compared to just using a plain string and relying on conversion?)

Reply:
It's not entirely clear what this does -- create and run a lambda that contains a const char* with the identifier, and then construct and move-return a new RTLIL::IdString? -- and what advantage this has over the implicit construction of an RTLIL::IdString which we had before?

Reply:
AIUI, the purpose is the created IdString is `static const`, so it will only be constructed (with associated expensive hash table lookup) once at startup rather than every time it is used.

Reply:
Ooh yes. I missed that. 

I think I initially discarded it because I was hoping this PR would be guaranteed to share the same IdString object between all users of each ID, but now I'm beginning to see that it only helps multiple constructions of the same ID for each particular usage location, and maybe a smart compiler would be able to identify that it can merge multiple `static const` IDs together. 

I am concerned that there is no _guarantee_ that the lambda will be inlined and executed just once though.

> so it will only be constructed (with associated expensive hash table lookup) once at startup rather than every time it is used.

My understanding is that it will only be executed upon the first invocation, after which it will not construct again. 

Also, wouldn't keeping all these statics around increase memory usage, since every appearance of each identifier could be duplicated?

Reply:
The memory requirements for the static IdStrings should be minimal. It's one `int` for the IdString and one `bool` the compiler adds to guard the static initialization code.

Another advantage of the `ID(A)` and `ID($add)` syntax over the current coding style is that it is much easier to search-and-replace when/if we are creating something like dedicated pre-allocated `ID_A` and `ID_add` later.

My guess would be that with `ID()`, reference counting has a bigger performance impact than ID allocation. So imo the main advantage of something like `ID_A` would be that we could build it in a way so that we don't need to reference count those IDs. (I also have on my todo an ifdef hack that disables reference counting, so we can benchmark that to get a baseline figure for how much overhead that really creates.)

I'd also guess that the overhead of having to allocate IdString objects the first time a code path is executed is neglectable. The bigger issue is that it has to check every time if this is the first run.

Reply:
All sounds good. It wasn't clear if this whether this was the complete solution, so maybe I shouldn't be critiquing it with the complete bar.